### PR TITLE
[wip] Use pagination with DescribeChangeset API

### PR DESCRIPTION
**Pull request by arunasank**
Fixes https://github.com/mapbox/cfn-config/issues/151

### CODE REVIEW

**Comment by arunasank** [comment]
Created a sample template with 150 SNS topics:
The names of the topics range from Topic0 - Topic149, and you can see that all 150 show up.
```
Action  Name      Type             Replace
------  --------  ---------------  -------
Modify  Topic0    AWS::SNS::Topic  true
Modify  Topic1    AWS::SNS::Topic  true
Modify  Topic10   AWS::SNS::Topic  true
Modify  Topic100  AWS::SNS::Topic  true
Modify  Topic101  AWS::SNS::Topic  true
Modify  Topic102  AWS::SNS::Topic  true
Modify  Topic103  AWS::SNS::Topic  true
Modify  Topic104  AWS::SNS::Topic  true
Modify  Topic105  AWS::SNS::Topic  true
Modify  Topic106  AWS::SNS::Topic  true
Modify  Topic107  AWS::SNS::Topic  true
Modify  Topic108  AWS::SNS::Topic  true
Modify  Topic109  AWS::SNS::Topic  true
Modify  Topic11   AWS::SNS::Topic  true
Modify  Topic110  AWS::SNS::Topic  true
Modify  Topic111  AWS::SNS::Topic  true
Modify  Topic112  AWS::SNS::Topic  true
Modify  Topic113  AWS::SNS::Topic  true
Modify  Topic114  AWS::SNS::Topic  true
Modify  Topic115  AWS::SNS::Topic  true
Modify  Topic116  AWS::SNS::Topic  true
Modify  Topic117  AWS::SNS::Topic  true
Modify  Topic118  AWS::SNS::Topic  true
Modify  Topic119  AWS::SNS::Topic  true
Modify  Topic12   AWS::SNS::Topic  true
Modify  Topic120  AWS::SNS::Topic  true
Modify  Topic121  AWS::SNS::Topic  true
Modify  Topic122  AWS::SNS::Topic  true
Modify  Topic123  AWS::SNS::Topic  true
Modify  Topic124  AWS::SNS::Topic  true
Modify  Topic125  AWS::SNS::Topic  true
Modify  Topic126  AWS::SNS::Topic  true
Modify  Topic127  AWS::SNS::Topic  true
Modify  Topic128  AWS::SNS::Topic  true
Modify  Topic129  AWS::SNS::Topic  true
Modify  Topic13   AWS::SNS::Topic  true
Modify  Topic130  AWS::SNS::Topic  true
Modify  Topic131  AWS::SNS::Topic  true
Modify  Topic132  AWS::SNS::Topic  true
Modify  Topic133  AWS::SNS::Topic  true
Modify  Topic134  AWS::SNS::Topic  true
Modify  Topic135  AWS::SNS::Topic  true
Modify  Topic136  AWS::SNS::Topic  true
Modify  Topic137  AWS::SNS::Topic  true
Modify  Topic138  AWS::SNS::Topic  true
Modify  Topic139  AWS::SNS::Topic  true
Modify  Topic14   AWS::SNS::Topic  true
Modify  Topic140  AWS::SNS::Topic  true
Modify  Topic141  AWS::SNS::Topic  true
Modify  Topic142  AWS::SNS::Topic  true
Modify  Topic143  AWS::SNS::Topic  true
Modify  Topic144  AWS::SNS::Topic  true
Modify  Topic145  AWS::SNS::Topic  true
Modify  Topic146  AWS::SNS::Topic  true
Modify  Topic147  AWS::SNS::Topic  true
Modify  Topic148  AWS::SNS::Topic  true
Modify  Topic149  AWS::SNS::Topic  true
Modify  Topic15   AWS::SNS::Topic  true
Modify  Topic16   AWS::SNS::Topic  true
Modify  Topic17   AWS::SNS::Topic  true
Modify  Topic18   AWS::SNS::Topic  true
Modify  Topic19   AWS::SNS::Topic  true
Modify  Topic2    AWS::SNS::Topic  true
Modify  Topic20   AWS::SNS::Topic  true
Modify  Topic21   AWS::SNS::Topic  true
Modify  Topic22   AWS::SNS::Topic  true
Modify  Topic23   AWS::SNS::Topic  true
Modify  Topic24   AWS::SNS::Topic  true
Modify  Topic25   AWS::SNS::Topic  true
Modify  Topic26   AWS::SNS::Topic  true
Modify  Topic27   AWS::SNS::Topic  true
Modify  Topic28   AWS::SNS::Topic  true
Modify  Topic29   AWS::SNS::Topic  true
Modify  Topic3    AWS::SNS::Topic  true
Modify  Topic30   AWS::SNS::Topic  true
Modify  Topic31   AWS::SNS::Topic  true
Modify  Topic32   AWS::SNS::Topic  true
Modify  Topic33   AWS::SNS::Topic  true
Modify  Topic34   AWS::SNS::Topic  true
Modify  Topic35   AWS::SNS::Topic  true
Modify  Topic36   AWS::SNS::Topic  true
Modify  Topic37   AWS::SNS::Topic  true
Modify  Topic38   AWS::SNS::Topic  true
Modify  Topic39   AWS::SNS::Topic  true
Modify  Topic4    AWS::SNS::Topic  true
Modify  Topic40   AWS::SNS::Topic  true
Modify  Topic41   AWS::SNS::Topic  true
Modify  Topic42   AWS::SNS::Topic  true
Modify  Topic43   AWS::SNS::Topic  true
Modify  Topic44   AWS::SNS::Topic  true
Modify  Topic45   AWS::SNS::Topic  true
Modify  Topic46   AWS::SNS::Topic  true
Modify  Topic47   AWS::SNS::Topic  true
Modify  Topic48   AWS::SNS::Topic  true
Modify  Topic49   AWS::SNS::Topic  true
Modify  Topic5    AWS::SNS::Topic  true
Modify  Topic50   AWS::SNS::Topic  true
Modify  Topic51   AWS::SNS::Topic  true
Modify  Topic52   AWS::SNS::Topic  true
Modify  Topic53   AWS::SNS::Topic  true
```

**Comment by arunasank** [comment]
@freenerd and I ended up talking about this today, and realised there's more to the picture than my earlier test. This change has a strong dependence on the `data.Status` and the `data.NextToken` parameters. Further the changesets are not constrained by number of resource items, they're constrained by the 1MB size limit.

Going to take a second look at the code, and at writing some good tests to test out the various edge cases.

## TODO:
* [ ] Write tests for
  * [ ] Changesets with a set NextToken, and status CREATE_COMPLETE
  * [ ] Changesets without a NextToken and status CREATE_COMPLETE
  * [ ] Changesets with a NextToken and status FAILED
for starts.
* [ ] Need a good way to test this PR on a real stack, 150 SNS topics probably come nowhere close to the 1 MB limit.

**Comment by arunasank** [comment]
ok, the 1MB limit in the documentation is a mistake. The NextToken is part of the response object, when there are greater than 100 resources in the changeset. The Status, and the NextToken look something like:

```
CREATE_IN_PROGRESS undefined
CREATE_IN_PROGRESS undefined
CREATE_IN_PROGRESS undefined
CREATE_IN_PROGRESS undefined
CREATE_IN_PROGRESS undefined
CREATE_IN_PROGRESS undefined
CREATE_IN_PROGRESS <NextTokenValue>
CREATE_COMPLETE undefined
```

when trying to update a stack with 200 SNS resources. Significant observation is that the the status is still `CREATE_IN_PROGRESS` when there are pending resources to be retrieved, and once the first set of changes are received, more changes are expected to be received if the `NextToken` key is available.

**Comment by rclark** [comment]
I'd encourage you to explore the `.eachPage()` functionality exposed by the aws-sdk in order to avoid having to write your own pagination code.

File: lib/actions.js [hunk 1]
```diff
@@ -332,26 +332,27 @@ actions.ChangeSetNotExecutableError = error('ChangeSetNotExecutableError');
...
+      changesetDescriptions = data;
+      changesetDescriptions.Changes = changes;
+    } else {
+      changesetDescriptions = data;
+    }
+
+    if ((data.Status === 'CREATE_COMPLETE' || data.Status === 'FAILED' || data.status === 'DELETE_COMPLETE'))
+      return callback(null, changesetDescriptions);
+
+    done();
```
Comment (rclark) [inline lib/actions.js]: Something seems odd here. I believe the `.eachPage(callback)` function indicates that you've gotten all the pages by firing your callback function with `data === null`. It seems like you should `.eachPage()` until you get to that condition, and then after that you should `callback(null, changesetDescriptions)`. 

Does that make sense?

**Comment by arunasank** [comment]
@rclark, :wave: I was mistaken about the `eachPage` covering polling. Further, polling the function passed to the eachPage() event also doesn't seem to work for me. 

The state at https://github.com/mapbox/cfn-config/pull/156/commits/03357c3a5d5cb87d87e8a685b3972d2315394124 works and needs some linter fixes. Can you please take over this issue 🙏 

cc/ @taraadiseshan @kellyoung @vsmart too if any of you have some time. 🙏

**Comment by rclark** [comment]
I tried this out live, and with `.eachPage()` I got 1 page of results, while with manual pagination, I got 4 pages. Looks like `.eachPage()` doesn't work here.

File: lib/actions.js [hunk 2]
```diff
@@ -332,15 +332,26 @@ actions.ChangeSetNotExecutableError = error('ChangeSetNotExecutableError');
...
-function describeChangeset(cfn, name, changesetId, callback) {
-  cfn.describeChangeSet({ ChangeSetName: changesetId, StackName: name }, function(err, data) {
-    if (err) return callback(err);
-
-    if (data.Status === 'CREATE_COMPLETE' || data.Status === 'FAILED' || data.status === 'DELETE_COMPLETE')
-      return callback(null, data);
-
-    setTimeout(describeChangeset, 1000, cfn, name, changesetId, callback);
-  });
+function describeChangeset(cfn, name, changesetId, nextToken, callback) {
```
Comment (freenerd) [inline lib/actions.js]: The `nextToken` can be removed here, since it's only needed when using `callAPI` below

File: lib/actions.js [hunk 3]
```diff
@@ -332,15 +332,26 @@ actions.ChangeSetNotExecutableError = error('ChangeSetNotExecutableError');
...
+    cfn.describeChangeSet({ ChangeSetName: changesetId, StackName: name, NextToken: nextToken }, function(err, data) {
+      if (err) return callback(err);
+
+      if (changesetDescriptions) {
+        var changes = [].concat(changesetDescriptions.Changes, data.Changes);
+        changesetDescriptions = data;
+        changesetDescriptions.Changes = changes;
+      } else {
+        changesetDescriptions = data;
+      }
```
Comment (freenerd) [inline lib/actions.js]: the `if` could be removed

```js
var changesetDescriptions = { Changes: [] };
...
// save latest data in changesetDescriptions but retain paginated Changes
var changes = [].concat(changesetDescriptions.Changes, data.Changes || []);
changesetDescriptions = data;
changesetDescriptions.Changes = changes;
```

File: test/actions.test.js [hunk 4]
```diff
@@ -489,20 +505,7 @@ test('[actions.diff] success', function(assert) {
...
-        {
-          action: 'Modify',
-          name: 'Topic',
-          id: 'arn:aws:sns:us-east-1:123456789012:my-stack-Topic-DQ8MBRPFONMK',
-          type: 'AWS::SNS::Topic',
-          replacement: false
-        }
-      ]
-    }, 'returned changeset details');
+    assert.deepEqual(data, { id: 'aa507e2bdfc55947035a07271e75384efe', status: 'CREATE_COMPLETE', execution: 'AVAILABLE', changes: [{ id: 'arn:aws:sns:us-east-1:123456789012:another-stack-Topic-ABCDEFGHIJKL', name: 'Topic', type: 'AWS::SNS::Topic', action: 'Modify', replacement: false }, { id: 'arn:aws:sns:us-east-1:123456789012:my-stack-Topic-DQ8MBRPFONMK', name: 'Topic', type: 'AWS::SNS::Topic', action: 'Modify', replacement: false }] }, 'returned changeset details');
```
Comment (freenerd) [inline test/actions.test.js]: i think `assert.equal('polled', 4)` would be good to test here as well

File: test/actions.test.js [hunk 5]
```diff
@@ -447,17 +447,33 @@ test('[actions.diff] success', function(assert) {
...
+        Type: 'Resource',
+        ResourceChange: {
+          Action: 'Modify',
+          LogicalResourceId: 'Topic',
+          PhysicalResourceId: 'arn:aws:sns:us-east-1:123456789012:another-stack-Topic-ABCDEFGHIJKL',
+          ResourceType: 'AWS::SNS::Topic',
+          Replacement: 'False'
+        }
+      }]
+    });
```
Comment (freenerd) [inline test/actions.test.js]: i'd love to see two more response types:

- [ ] having no .`Changes` returned
- [ ] having a `NextToken` returned
